hazelcast-jet-core/src/test/java/com/hazelcast/jet/core/SnapshotFailureTest.java

```diff
@@ -97,28 +103,25 @@ public void when_snapshotFails_then_jobShouldNotFail() throws Exception {
 
         Job job = instance1.newJob(dag, config);
 
-        IMap<Object, Object> snapshotsMap = instance1.getMap(snapshotsMapName(job.getId()));
-
-        SnapshotRecord[] failedRecord = new SnapshotRecord[1];
-        while (failedRecord[0] == null && !job.getFuture().isDone()) {
-            // find a failed record in snapshots map
-            snapshotsMap
-                    .values().stream()
-                    .filter(r -> r instanceof SnapshotRecord)
-                    .map(r -> (SnapshotRecord) r)
-                    .filter(r -> r.status() == SnapshotStatus.FAILED)
-                    .findFirst()
-                    .ifPresent(r -> failedRecord[0] = r);
-            LockSupport.parkNanos(MILLISECONDS.toNanos(1));
-        }
+        // let's start a second job that will watch the snapshots map and write failed
+        // SnapshotRecords to a list, which we will check for presence of failed snapshot
+        Pipeline p = Pipeline.create();
+        p.drawFrom(Sources.mapJournal(snapshotsMapName(job.getId()),
+                event -> event.getNewValue() instanceof SnapshotRecord
+                        && ((SnapshotRecord) event.getNewValue()).status() == SnapshotStatus.FAILED,
+                EventJournalMapEvent::getNewValue,
+                JournalInitialPosition.START_FROM_OLDEST))
+         .peek()
+         .drainTo(Sinks.list("failed_snapshot_records"));
+        instance1.newJob(p);
 
         job.join();
```

Comment:
Could it be happen that before second job is written it's results to the list, first job finishes and make the following assertions earlier and fails ?

Reply:
The first job takes 5 seconds to complete, it does snapshot every 100ms. In this case there were 6 failed snapshots made. If the second job is not able to process any event from the journal during 5 seconds, it will fail. If it ever fails, i'll increase the limits.

Reply:
👍